### PR TITLE
Improve adjoint for product and zip

### PR DESCRIPTION
**Pull request by lxvm**
Hi,

I've returned to my first contribution in #1170 since I noticed I couldn't differentiate w.r.t `Iterator.product`s that have a number as an iterator. This pr adds a test and fixes the issue while also improving the inferrability of the adjoint.

### PR Checklist

- [x] Tests are added
- [x] Documentation, if applicable

### CODE REVIEW

File: src/lib/array.jl [hunk 1]
```diff
@@ -282,24 +282,32 @@ end
 _ndims(::Base.HasShape{d}) where {d} = d
 _ndims(x) = Base.IteratorSize(x) isa Base.HasShape ? _ndims(Base.IteratorSize(x)) : 1
 
+function prodfunc(xs, dy)
+  @assert length(first(dy)) == length(xs)
+  ndim = map(Zygote._ndims, xs)
+  cdim = cumsum((0, ndim[begin:end-1]...))
+  getters = ntuple(n -> StaticGetter{n}(), Val(length(xs)))
+  dims = Vector{Int}(undef, length(xs))
```
[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/lib/array.jl:290]: Is the problem that the old `nd = _ndims(xs[n])` failed to const prop? If so, then this vector approach makes sense to me.

[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/lib/array.jl]: Great that you got the tuple path working!

File: src/lib/array.jl [hunk 2]
```diff
@@ -268,24 +268,32 @@ end
...
-      return _project(xs[n], reshape(red, axes(xs[n])))
-    end
-  end
+  back(dy::AbstractArray) = prodfunc(xs, dy)
   Iterators.product(xs...), back
 end
 
+@adjoint function Base.collect(p::Base.Iterators.ProductIterator)
+  collect(p), dy -> ((iterators=prodfunc(p.iterators, dy),),)
+end
```
Comment (ToucheSir) [inline src/lib/array.jl:295]:
```suggestion
@adjoint function Base.collect(p::Base.Iterators.ProductIterator)
  collect_product_pullback(dy) = ((iterators=prodfunc(p.iterators, dy),),)
  return collect(p), collect_product_pullback
end
```
Naming pullbacks makes debugging and stacktraces much nicer :)

**Review by ToucheSir — COMMENTED** [review]
This mostly LGTM. Are you able to add tests which check that one or more of the gradient computation, pullback or `productfunc` are type stable? With the effort put into this, we should make sure inference keeps working!

**Comment by ToucheSir** [comment]
Although it does use ChainRules' projection machinery sometimes, Zygote overall doesn't do projection quite the same way because it predates ChainRules. The legacy projection machinery we do have can be rather inconsistent. In this particular case however, it looks like the input type is causing null gradients?
```
julia> gradient(x -> sum(prod, zip(x)), collect(1:5))
([1.0, 1.0, 1.0, 1.0, 1.0],)

julia> gradient(x -> sum(prod, zip(x)), 1.0:5.0)
([1.0, 1.0, 1.0, 1.0, 1.0],)
```
It makes sense that an integer range would be considered non-differentiable, but it would be good to confirm Zygote is doing this for the right reason and not because of some bug. Either way, if you can't figure out `zip` easily I'd just leave it for a follow-up PR and we can try to get the `product` changes in first.

**Comment by ToucheSir** [comment]
It turns out the answer is easier than I thought: https://github.com/FluxML/Zygote.jl/blob/54f1e807d5c098a6100c422b54d14f8cd85e0b3c/src/lib/array.jl#L252. This may have been for supporting `for x in 1:N ...`.

**Comment by mcabbott** [comment]
Zygote often throws away gradients of a UnitRange, here:

https://github.com/FluxML/Zygote.jl/blob/54f1e807d5c098a6100c422b54d14f8cd85e0b3c/src/lib/array.jl#L252

It's not enforced by projection, so things that hit other rules such as `gradient(x -> sum(abs2, x), 1:5)` don't give nothing.

> I did switch the adjoint for the constructor Iterators.Zip to one for the function Iterators.zip and I'm not sure if something depended especially on the former since there were previously no tests for it.

I have no memory of why, but when initially writing these rules, attaching them to the uppercase constructor not the lowercase function somehow made more cases work. There are tests [here](https://github.com/FluxML/Zygote.jl/blob/54f1e807d5c098a6100c422b54d14f8cd85e0b3c/test/features.jl#L619) but fewer than I thought.

File: src/lib/array.jl [hunk 3]
```diff
@@ -169,8 +170,11 @@ _reverse(x::Symmetric) = Symmetric(_reverse(x.data), x.uplo == 'U' ? :L : :U)
 # So we keep axes(x) to restore gradient dx to its full length & correct shape.
 _tryaxes(x) = axes(x)
 _tryaxes(x::Tuple) = Val(length(x))
-_restore(dx, ax::Tuple) = axes(dx) == ax ? dx : reshape(vcat(dx, falses(prod(length, ax) - length(dx))), ax)
+_tryaxes(::Number) = Val(-1)
```
Comment (ToucheSir) [inline src/lib/array.jl:173]: Instead of a `Val`, maybe `nothing` or `missing` would be a more appropriate sentinel value here?

**Review by ToucheSir — APPROVED** [review]
LGTM, a couple of things to touch up before merging.

**Comment by ToucheSir** [comment]
This is a great contribution for a tricky set of rules, thanks @lxvm !

**Pull request merged**